Extract code from content tagger to parse a taxonomy into a navigable tree structure

MatMoore opened the pull request:
Code migrated from content tagger (`TreeNode`) via Whitehall.

This should allow us to simplify a lot of code in
- Whitehall (https://github.com/alphagov/whitehall/pull/2997)
- Rummager (https://github.com/alphagov/rummager/pull/748)
- Content tagger (https://trello.com/c/ykmIjJI5/484-use-the-new-publishing-api-expanded-links-feature-in-content-tagger)

Changes made:

- Removed dependencies on ActiveSupport
- Added `#descendants`, `#ancestors`, `#breadcrumb_trail` methods
- Renamed `TreeNode` to `LinkedEdition` and `#node_depth` to `#depth` 
- Added `#add_taxons` and `#taxons` that allow regular content items to be included in the tree structure.
- Added `#taxons_with_ancestors` to go from a piece of content to all taxons that are affected by it. This is what we're calling `part_of_taxonomy` in rummager: https://github.com/alphagov/rummager/pull/748
- Added custom `#inspect` method for debuggability.
-  Added a parser class to build the tree from the publishing api expanded links response. This now expands `parent_taxons` and `child_taxons`.
- Don't store additional state in a content item hash any more. This couples us to the way publishing api returns stuff and makes it more complicated to parse nested items, which have a slightly different format.

Commit:
Extract code from content tagger

Code migrated from content tagger (`TreeNode`) via Whitehall.

See also https://github.com/alphagov/whitehall/pull/2997

Commit:
Separate response parsing from tree traversal

MatMoore committed:
Parse parent_taxons and taxons links

This allows you traverse the taxonomy in both directions from
a starting taxon.

It also lets you get the taxons related to an edition.

When a document is published, it affects:
- pages for directly tagged taxons
- pages for the parents of directly taxons (and their parents and so on)
- searches and subscriptions scoped to any of these taxons

MatMoore committed:
Remove unused helper

MatMoore commented:
I've expanded the functionality to follow expansions of parent_taxons as well as child_taxons, so that we can handle breadcrumbs and things in publishing apps.

I previously attempted this in https://github.com/alphagov/govuk_navigation_helpers/pull/26 but we decided that wasn't the best place for it. (cc @carvil)

I still need to set up CI for this gem but I think I'll do that in a separate branch.

tijmenb commented on README.md:
```diff
@@ -20,26 +20,52 @@ Or install it yourself as:
-content_store = GdsApi::ContentStore.new(Plek.new.find('content-store'))
-content_item = content_store.content_item('/education')
-taxonomy = GovukTaxonomyHelpers::Taxon.new(content_item)
+publishing_api = GdsApi::PublishingApi.new(Plek.new.find('publishing-api'))
```
By using `Services.publishing_api` you can omit much of the plek/gds-api boilerplate.

MatMoore replied:
Fair enough. I added it so it's copy and pastable into a console but `Services.publishing_api` should be fine too.

tijmenb commented on README.md:
````diff
@@ -20,26 +20,52 @@ Or install it yourself as:
-puts taxonomy.descendants(&:name) # All descendant taxons
-puts taxonomy.children.map(&:name) # All direct children
+```ruby
+puts taxonomy.tree             # All taxons in this branch of the taxonomy
````
Perhaps pasting the outputs here would be good. Alternatively, referring to the method docs.

MatMoore replied:
I still need to actually write the docs, so might do this all at once in a follow up PR.

tijmenb commented on README.md:
````diff
@@ -20,26 +20,52 @@ Or install it yourself as:
+)
+```
+
+A `LinkedEdition` built from a taxon can access all *narrower term* taxons below it and all *broader term* taxons above it.
````
Why `Edition`?

MatMoore replied:
That's what publishing api calls it now, and I think it's more descriptive (vs `content item`)

tijmenb replied:
I'm really unsure about muddling the terminology. Why are we asking for a `content_item` above? Is the word "content item" no longer a thing?

MatMoore replied:
Mixing the two was an oversight, I'll make it consistent.

According to https://github.com/alphagov/publishing-api:

> Edition: The content of a document is represented by an edition, it represents a distinct version of a Document.
> Content Item: A representation of content that can be sent to a content store.

tijmenb replied:
Just had a enlightening conversation with @kevindew. He helpfully pointed out that the thing returned from `publishing_api.get_content` is neither a content item (which would includes `links`) nor a edition (which doesn't include things like a `content_id`). Edition is still an implementation concern of publishing-api.

There's going to be conversations soon about publishing-api's API & terminology, so let's chat about this tomorrow and come up with something.

MatMoore replied:
Discussed offline: we'll use content item for now and revisit later if we get to more of a consensus on terminology.

tijmenb commented on README.md:
```diff
@@ -20,26 +20,52 @@ Or install it yourself as:
+content_item = publishing_api.get_content("c75c541-403f-4cb1-9b34-4ddde816a80d")
+expanded_links = publishing_api.get_expanded_links("c75c541-403f-4cb1-9b34-4ddde816a80d")
+
+taxonomy = GovukTaxonomyHelpers.parse_publishing_api_response(
```
I maybe don't associate this with parsing. What would you think of something like this?

```ruby
GovukTaxonomyHelpers::WhateverObjectYouGetBack.new(
  content_item: content_item,
  expanded_links: expanded_links
)
```

MatMoore replied:
The thing you get back is `LinkedEdition` (formally known as `TreeNode`), which you can also create like this 

```ruby
foo = GovukTaxonomyHelpers::LinkedEdition.new(
  name: "foo",
  base_path: "/dsfsdf/sdffsd",
  content_id: "abc"
)
```

and you add children to it with `<<` (there are better examples in the spec, and I'll document this separately).

`#parse_publishing_api_response` is a helper function around the `PublishingApiResponseParser` class, which itself returns the `LinkedEdition`. 

I wanted to separate these two things, because you shouldn't have to call the publishing api to use `LinkedEdition`, and the response handling code is likely to change more quickly than the `LinkedEdition` itself, which is basically just a fancy tree structure.

We already do similar taxonomy traversal things on the frontend, using the content store response, and while that looks very similar to the publishing api response, there is no guarantee that it will stay that way. I've focused just on publishing apps in this PR, but I want to make it possible to extend to the content store as well later if that turns out to be useful.

Maybe `parse` is not the best name, but I think it needs to be obvious that the arguments it takes are expected to come from the publishing api, because it's likely going to break if passed anything else.

tijmenb replied:
`GovukTaxonomyHelpers::LinkedEdition.from_publishing_api` or something?

tijmenb reviewed (changes requested):
Cool! I've mostly commented on the docs/API because we can always change the implementation.

MatMoore committed:
Rename the public API to be clearer

`Edition` follows the publishing api naming conventions:
https://github.com/alphagov/publishing-api

`Content item` means more or less the same thing but also specifically the
expanded representation (with links) that is sent to the content store.

We're preferring "content item" here until there is some sort of agreement on
what words mean.

I've removed `parse_publishing_api_response` since it was unclear what this returned.

MatMoore committed:
Simplify public API

Add LinkedContentItem.from_publishing_api method so that you don't
have to deal with PublishingApiResponse classes directly.

MatMoore commented:
@tijmenb I've made the suggested API changes. If it's ok to merge now, I'll raise a separate PR for docs and CI.

MatMoore committed:
Add yardoc documentation

Add method level documentation and only include basic examples in the README.

MatMoore commented:
Added docs here: https://github.com/alphagov/govuk_taxonomy_helpers/pull/3

Commit:
Merge pull request #3 from alphagov/docs

Add yardoc documentation

MatMoore commented:
@tijmenb are you happy with the changes?
